### PR TITLE
Add: candidate's pledge and thawing of token

### PR DESCRIPTION
**Pull request by xhuangdx**
### Description
1. Become a candidate need to mortgage token

2. Cancel the candidate can unlock the token, unlock a certain lock period

3. Become a candidate for the threshold, determine the balance to meet the conditions to successfully apply

4. If the user customizes the RewardRatio, then the RewardRatio must be within 100%.

### Document 

https://github.com/DxChainNetwork/engdoc/blob/master/project/godx/design_doc/dpos_candidate/candidate.md

### Issue

https://docs.google.com/document/d/1IVQVjrsBBmVVOfv699tkBicUcpZM-N2UNjmHsfAy378/edit

### Type of change
- [x] New Feature

### Test case
core/vm/evm_test.go:50

### CODE REVIEW

File: core/vm/evm.go [hunk 1]
```diff
@@ -204,6 +204,19 @@ func (evm *EVM) Call(caller ContractRef, addr common.Address, input []byte, gas
 		return nil, gas, ErrInsufficientBalance
 	}
 
+	//The pledge asset of candidate will not be used.
+	if fa := evm.StateDB.GetState(caller.Address(), common.BytesToHash([]byte("FrozenAssets"))); fa != (common.Hash{}) {
+		balance := evm.StateDB.GetBalance(caller.Address())
+		if value.Cmp(new(big.Int).Sub(balance, fa.Big())) < 0 {
```
Comment (xfliudx) [inline core/vm/evm.go]: value > balance - fa.Big()

Comment (xhuangdx) [reply core/vm/evm.go]: Thanks for reminding, resolved

File: core/vm/evm.go [hunk 2]
```diff
@@ -518,6 +524,42 @@ func (evm *EVM) ApplyDposTransaction(txType string, dposContext *types.DposConte
...
+	err := dposContext.BecomeCandidate(caller)
+	if err != nil {
+		return nil, gas, err
+	}
+	evm.StateDB.SetState(caller, common.BytesToHash([]byte("FrozenAssets")), common.BigToHash(value))
+	// If the user customizes the RewardRatio, then the RewardRatio must be within 100%.
+	if len(data) != 0 {
+		rr := types.CandidateInfo{}
+		err := rlp.DecodeBytes(data, &rr)
+		if err == nil && rr.RewardRatio <= 100 {
```
[the remark this answers is not in the document]
Comment (xhuangdx) [reply core/vm/evm.go]: There is no need to return error here. If the user does not set this parameter, the system will use the default d....

File: core/types/dpos_context.go [hunk 3]
```diff
@@ -444,3 +444,8 @@ func (dc *DposContext) SetValidators(validators []common.Address) error {
 	dc.epochTrie.Update(key, validatorsRLP)
 	return nil
 }
+
+//CandidateInfo additional relevant data when sending the candidateTx.
```
[the remark this answers is not in the document]
Comment (xhuangdx) [reply core/types/dpos_context.go]: Currently discussed with Xiang Zhang, temporarily placed under the dpos_context.go file, candidates also need to send the candidate transaction, so also need to pledge

File: core/vm/evm.go [hunk 4]
```diff
@@ -518,6 +524,42 @@ func (evm *EVM) ApplyDposTransaction(txType string, dposContext *types.DposConte
...
+	return nil, gas, nil
+}
+
+// CandidateCancelTx cancellation of candidate thawing assets requires a defrosting period.
+func (evm *EVM) CandidateCancelTx(caller common.Address, data []byte, gas uint64, dposContext *types.DposContext) ([]byte, uint64, error) {
+	err := dposContext.KickoutCandidate(caller)
+	if err != nil {
+		return nil, gas, err
+	}
+	tp := common.BigToHash(new(big.Int).SetUint64(evm.Context.BlockNumber.Uint64() + params.ThawingPeriod))
```
[the remark this answers is not in the document]
Comment (xhuangdx) [reply core/vm/evm.go]: This Xiang Zhang has completed, which can be seen on his pull quest.

**Pull request closed without merging**